airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py

```diff
@@ -0,0 +1,240 @@
+        self.tools: BulkTools = BulkTools()
+
+    # 5Mb chunk size to save the file
+    retrieve_chunk_size = 1024 * 1024 * 5
+
+    # time between job status checks
+    job_check_interval_sec: int = 5
+
+    # max attempts for job creation
+    concurrent_max_retry = 10
+    # attempt counter
+    concurrent_attempt = 0
+    # attempt limit indicator
+    concurrent_max_attempt_reached = False
+    # sleep time per creation attempt
+    concurrent_interval_sec = 45
+
+    def get_errors_from_response(self, response: requests.Response) -> Iterable[Mapping[str, Any]]:
+        try:
+            return response.json().get("errors") or response.json().get("data", {}).get("bulkOperationRunQuery", {}).get("userErrors", [])
+        except (Exception, JSONDecodeError) as e:
+            raise ShopifyBulkExceptions.BulkJobBadResponse(
+                f"Couldn't check the `response` for `errors`, response: `{response.text}`. Trace: {repr(e)}."
+            )
+
+    def has_running_concurrent_job(self, response: requests.Response) -> bool:
+        """
+        When concurent BULK Job is already running for the same SHOP we receive:
+        Error example:
+        [
+            {
+                'field': None,
+                'message': 'A bulk query operation for this app and shop is already in progress: gid://shopify/BulkOperation/4039184154813.',
+            }
+        ]
+        """
+
+        concurent_job_pattern = "A bulk query operation for this app and shop is already in progress"
+        errors = self.get_errors_from_response(response)
+        if errors:
+            for error in errors:
+                message = error.get("message", "")
+                if concurent_job_pattern in message:
+                    return True
+                else:
+                    continue
+        else:
+            return False
+
+    @limiter.balance_rate_limit(api_type=ApiTypeEnum.graphql.value)
+    def job_check_for_errors(self, response: requests.Response) -> None:
+        try:
+            # checking for other errors
+            errors = self.get_errors_from_response(response)
+            if errors:
+                # when we have the scenario we didn't face yet.
+                raise ShopifyBulkExceptions.BulkJobError(f"Something wong with the BULK job, errors: {errors}")
+        except JSONDecodeError as e:
+            raise ShopifyBulkExceptions.BulkJobBadResponse(
+                f"The BULK Job result contains errors or could not be parsed. Details: {response.text}. Trace: {e}"
+            )
+
+    def job_get_id(self, response: Optional[requests.Response] = None) -> Optional[str]:
+        response_data = response.json() if response else {}
+        bulk_response = response_data.get("data", {}).get("bulkOperationRunQuery", {}).get("bulkOperation", {})
+        if bulk_response and bulk_response.get("status") == ShopifyBulkStatus.CREATED.value:
+            job_id = bulk_response.get("id")
+            if job_id:
```

Comment:
when would a None job_id be valid? should we raise an exception if that happens?

Reply:
corrected.